test/common/config/datasource_test.cc

```diff
@@ -384,89 +389,105 @@ TEST_F(AsyncDataSourceTest, LoadRemoteDataSourceExpectInvalidData) {
+  EXPECT_EQ(async_data, body);
+}
 
-  {
-    AsyncDataSourcePb config;
+TEST_F(AsyncDataSourceTest, BaseIntervalGreaterThanMaxInterval) {
+  AsyncDataSourcePb config;
 
-    std::string yaml = R"EOF(
+  std::string yaml = R"EOF(
     remote:
       http_uri:
         uri: https://example.com/data
         cluster: cluster_1
       sha256:
         b94d27b9934d3e08a52e52d7da7dabfac484efe37a5380ee9088f7ace2efcde9
+      retry_policy:
+        retry_back_off:
+          base_interval: 10s
+          max_interval: 1s
+        num_retries: 3
   )EOF";
-    TestUtility::loadFromYaml(yaml, config);
-    EXPECT_TRUE(config.has_remote());
-
-    EXPECT_CALL(cm_, httpAsyncClientForCluster("cluster_1")).WillOnce(ReturnRef(cm_.async_client_));
-    EXPECT_CALL(cm_.async_client_, send_(_, _, _))
-        .WillOnce(
-            Invoke([&](Http::RequestMessagePtr&, Http::AsyncClient::Callbacks& callbacks,
-                       const Http::AsyncClient::RequestOptions&) -> Http::AsyncClient::Request* {
-              Http::ResponseMessagePtr response(
-                  new Http::ResponseMessageImpl(Http::ResponseHeaderMapPtr{
-                      new Http::TestResponseHeaderMapImpl{{":status", "200"}}}));
-              response->body() = std::make_unique<Buffer::OwnedImpl>(body);
-
-              callbacks.onSuccess(std::move(response));
-              return nullptr;
-            }));
+  TestUtility::loadFromYaml(yaml, config);
+  EXPECT_TRUE(config.has_remote());
 
-    EXPECT_CALL(init_manager_, add(_)).WillOnce(Invoke([this](const Init::Target& target) {
-      init_target_handle_ = target.createHandle("test");
-    }));
+  EXPECT_THROW_WITH_MESSAGE(std::make_unique<Config::DataSource::RemoteAsyncDataProvider>(
+                                cm_, init_manager_, config.remote(), dispatcher_, random_, true,
+                                [&](const std::string&) {}),
+                            EnvoyException,
+                            "max_interval must be greater than or equal to the base_interval");
+}
 
-    provider = std::make_unique<Config::DataSource::RemoteAsyncDataProvider>(
-        cm_, init_manager_, config.remote(), true, [&](const std::string& data) {
-          EXPECT_EQ(init_manager_.state(), Init::Manager::State::Initializing);
-          EXPECT_EQ(data, body);
-          async_data = data;
-        });
-  }
+TEST_F(AsyncDataSourceTest, BaseIntervalTest) {
+  AsyncDataSourcePb config;
 
-  EXPECT_CALL(init_manager_, state()).WillOnce(Return(Init::Manager::State::Initializing));
-  EXPECT_CALL(init_watcher_, ready());
-  init_target_handle_->initialize(init_watcher_);
-  EXPECT_EQ(async_data, body);
-  EXPECT_NE(nullptr, provider.get());
+  std::string yaml = R"EOF(
+    remote:
+      http_uri:
+        uri: https://example.com/data
+        cluster: cluster_1
+      sha256:
+        xxx
+      retry_policy:
+        retry_back_off:
+          base_interval: 0.0001s
```

Comment:
It seems the enforcement in the proto file doesn't work. @mattklein123 Am I missing sth. here?

Reply:
I expect the test to fail here.

Reply:
Validation is explicit. Use `loadFromYamlAndValidate` in this file. In the real server this would be done recursively for you.